### PR TITLE
Cluster default publishing strategy for cluster operators

### PR DESCRIPTION
**Pull request by abhinavdahiya**
/cc @openshift/api-reviewers @openshift/openshift-team-installer @sdodson @ironcladlou

### CODE REVIEW

File: cluster-default-publishing-strategy.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+Modify the existing `config.openshift.io`'s custom resource [`Infrastructure`][infrastructure-api] to include an enum field `Publish` that can be set to `External` or `Internal`.
+
+The field will be added to [`status` section][infrastructure-api-status] because the goal is to provide an API for discovery i.e. the current state, and not to provide an API for customers to specify the desired state.
+
+### User Stories
+
+#### Default Ingress Controller LoadBalancer Scope
+
+The ingress-operator will be able to use this field to define the **default** [scope][ingress-controller-api-load-balancer-scope] for the LoadBalancers for the [ingress controllers][ingress-controller-api].
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: @ironcladlou is this related to your designs?

File: cluster-default-publishing-strategy.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+Publish PublishingStrategy `json:”publish,omitempty”`
+
+...
+
+}
+```
+
+### Migration of existing clusters
+
+All clusters with `empty` PublishingStrategy can be updated to `External`, while all non-empty values will be kept untouched.
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: Which operator/process actually does this?

File: cluster-default-publishing-strategy.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+#### Rollbacks
+
+In case of rollbacks after the migration has been done, the value `External` will be accurate.
+
+### Test Plan
+
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: how would this be tested?

File: cluster-default-publishing-strategy.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+The API defines `empty` equivalent to `External`, therefore all the clients will have to keep that check to remain backwards compatible.
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+- Because the ingres-operator is currently the only user of this API field, the `PublishingStrategy` could also be added to the [`Ingress`][ingress-api] custom resource. But because the strategy affects more than the ingress, like the
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: typo: ingres

File: cluster-default-publishing-strategy.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+Network security filters on clouds, this field is more suited for `Infrastructure` custom resource in the long term.
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: @ironcladlou claim here.

**Comment by deads2k** [comment]
@derekwaynecarr I think we have a need to expand the template to include something like: "impacted operators"

**Comment by smarterclayton** [comment]
I don’t see anything objectionable.  Will defer to David but this has my approval at a high level (barring usual squabbling at merge time obviously)

File: cluster-default-publishing-strategy.md (hunk 1 above)
Comment (ironcladlou) [reply cluster-default-publishing-strategy.md]: https://github.com/openshift/enhancements/pull/31 is "bring your own IngressController", and this PR is (I think) about adding some new focused API so that you don't have to bring your own (usually, and for this specific use case).

File: cluster-default-publishing-strategy.md (hunk 5 above)
Comment (ironcladlou) [reply cluster-default-publishing-strategy.md]: The only use cases I know of so far related to this API would. be the installer-created API LB, and the ingress LB. Not sure about anything else.

File: cluster-default-publishing-strategy.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Addition of the new field requires migration of all the existing clusters. Since we have only supported `External` strategy for OpenShift, the migration should be simple enough.
+
+## Design Details
+
+### New API
```
Comment (ironcladlou) [inline cluster-default-publishing-strategy.md]: Would this be a reasonable backport candidate?

File: cluster-default-publishing-strategy.md (hunk 3 above)
Comment (abhinavdahiya) [reply cluster-default-publishing-strategy.md]: @deads2k do you mean?

a) test if this API is set correctly?
b) Or if the migration is happening correctly?

File: cluster-default-publishing-strategy.md (hunk 6 above)
Comment (abhinavdahiya) [reply cluster-default-publishing-strategy.md]: The new API is backwards compatible because it only changes the behavior when set. the default today is External and not set is also External. only when its set to `Internal` will the clients care.

Also the `Migration of existing clusters` talks about quick and easy migration.

Comment (ironcladlou) [reply cluster-default-publishing-strategy.md]: I'm talking about backporting the API and functionality, once available, to 4.1 or 4.2.

File: cluster-default-publishing-strategy.md (hunk 5 above)
Comment (abhinavdahiya) [reply cluster-default-publishing-strategy.md]: LBs require more than LB itself, it requires the network firewall rules are setup correctly. We need this information in API to make sure we can transfer a lot of the installer work into the bootstrap-host and controllers that manage these, because we don't today.

File: cluster-default-publishing-strategy.md (hunk 6 above)
Comment (ironcladlou) [reply cluster-default-publishing-strategy.md]: I meant 4.2, which is the first release that supports internal IngressControllers.

Comment (abhinavdahiya) [reply cluster-default-publishing-strategy.md]: I don't think we back port API changes.

Comment (ironcladlou) [reply cluster-default-publishing-strategy.md]: @smarterclayton if there was sufficient demand, do you see any technical obstacles to a backport?

Basically what I'm trying to figure out is, given we accept this proposal generally (and I do), does that mean we _shouldn't_ do https://github.com/openshift/enhancements/pull/25 and should instead continue addressing more specific use cases as they arise? If we don't do https://github.com/openshift/enhancements/pull/25 and we can't backport this proposed API, we have no supportable solution for 4.2 users. Just something to consider.

File: cluster-default-publishing-strategy.md (hunk 3 above)
Comment (deads2k) [reply cluster-default-publishing-strategy.md]: > @deads2k do you mean?
> 
> a) test if this API is set correctly?
> b) Or if the migration is happening correctly?

This feature doesn't exist in a vacuum.  It's not really useful unless something is done with this field.  I'm asking how you would ensure that your initial consumers worked correctly.

**Comment by deads2k** [comment]
> LBs require more than LB itself, it requires the network firewall rules are setup correctly. We need this information in API to make sure we can transfer a lot of the installer work into the bootstrap-host and controllers that manage these, because we don't today.

Please add an enumerate list of examples.  If I've read, this correctly, I think it's,

1. ingress controller
2. bootstraphost - this doesn't get managed post 
    1. network firewall

What other pieces requires this knowledge and which ones persist after installation?

At a high-level, #31 appears to be a minimal, targeted solution where each individual piece is independently valuable.  I'm looking for a similarly compelling reason to expand our top level config.

File: cluster-default-publishing-strategy.md (hunk 6 above)
Comment (sdodson) [reply cluster-default-publishing-strategy.md]: @ironcladlou did you mean #31 rather than #25 here?

Comment (ironcladlou) [reply cluster-default-publishing-strategy.md]: @sdodson 
>@ironcladlou did you mean #31 rather than #25 here?

Oops, yes, edited, thanks!

**Comment by adambkaplan** [comment]
@deads2k +1 to impacted operators. Ex I don't think the registry operator is impacted by this proposal (our CRD lets admins create routes), but it could?

**Comment by patrickdillon** [comment]
> @derekwaynecarr I think we have a need to expand the template to include something like: "impacted operators"

>@deads2k +1 to impacted operators. Ex I don't think the registry operator is impacted by this proposal (our CRD lets admins create routes), but it could?

I suggest "affected operators" over "impacted operators."

File: cluster-default-publishing-strategy.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+last-updated: 2019-09-18
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Cluster Publishing Strategy
+
+Administrators would like to be able to install OpenShift cluster as `internal/private`, which is only accessible from within the network, VPN tunnels and, not visible to the Internet. Therefore all the cluster operators that create endpoints for the cluster would need a place to discover this intent of the administrator and ensure that their resources are only accessible from within the network, VPN tunnels and not visible to the Internet by default.
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: line breaks at about 120 chars so we can comment.

Installing is different than maintaining.  In the summary, you should describe the difference.

File: cluster-default-publishing-strategy.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+See the [section][#cluster-publishing-strategy]
+
+## Motivation
+
+See the [section][#cluster-publishing-strategy]
+
+### Goals
+
+1. Define an API that allows all the cluster operators to discover the default cluster publishing strategy.
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: do defend this goal, you need to list all the operators you believe are affected and which resources they are managing.

Comment (deads2k) [reply cluster-default-publishing-strategy.md]: This was a top level comment before and still outstanding. The only impact I know of is to the default ingress operator and nothing else.

File: cluster-default-publishing-strategy.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+### Non-Goals
+
+## Proposal
+
+Modify the existing `config.openshift.io`'s custom resource [`Infrastructure`][infrastructure-api] to include an enum field `Publish` that can be set to `External` or `Internal`.
```
Comment (deads2k) [inline cluster-default-publishing-strategy.md]: Based on an impact only to the default ingress operator, modifying infrastructure is hard to justify

**Comment by ironcladlou** [comment]
What if we move the field to `ingress.config.openshift.io` instead?

**Comment by ironcladlou** [comment]
>What if we move the field to ingress.config.openshift.io instead?

As described here: https://github.com/openshift/enhancements/pull/31/files#diff-40bb87d9f5690ccc8c6d75986d956f45R245

**Comment by sdodson** [comment]
There's agreement that this should not be in global cluster config. #31 has been accepted
/close
